File: typescript/sskr/src/spec.ts
import { MAX_SHARE_COUNT } from './constants.js';
import { SskrError } from './error.js';
import type { GroupSpec } from './group-spec.js';

/**
 * A specification for an SSKR split.
 */
export class Spec {
    readonly #groupThreshold: number;
    readonly #groups: GroupSpec[];

    private constructor(groupThreshold: number, groups: GroupSpec[]) {
        this.#groupThreshold = groupThreshold;
        this.#groups = groups;
    }

    /**
     * Creates a new `Spec` with the given group threshold and groups.
     *
     * @param groupThreshold - The minimum number of groups required to
     *   reconstruct the secret.
     * @param groups - The list of group specifications.
     * @returns A new `Spec` instance.
     * @throws {SskrError} If the group threshold is zero, greater than the
     *   number of groups, or if the number of groups exceeds
     *   {@link MAX_SHARE_COUNT}.
     */
    static create(groupThreshold: number, groups: GroupSpec[]): Spec {
        if (groupThreshold === 0) {
            throw SskrError.groupThresholdInvalid();
        }
        if (groupThreshold > groups.length) {
            throw SskrError.groupThresholdInvalid();
        }
        if (groups.length > MAX_SHARE_COUNT) {
            throw SskrError.groupCountInvalid();
        }
        return new Spec(groupThreshold, [...groups]);
    }

    /** The group threshold. */
    get groupThreshold(): number {
        return this.#groupThreshold;
    }

    /** The group specifications. */
    get groups(): readonly GroupSpec[] {
        return this.#groups;
    }

    /** The number of groups. */
    get groupCount(): number {
        return this.#groups.length;
    }

    /** The total number of shares across all groups. */
    get shareCount(): number {
        return this.#groups.reduce((sum, g) => sum + g.memberCount, 0);
    }
}

File: go/sskr/spec.go
package sskr

import (
	"fmt"
	"strconv"
	"strings"
)

// Spec is a specification for an SSKR split.
type Spec struct {
	groupThreshold int
	groups         []GroupSpec
}

// NewSpec creates a new Spec with the given group threshold and groups.
func NewSpec(groupThreshold int, groups []GroupSpec) (Spec, error) {
	if groupThreshold == 0 {
		return Spec{}, ErrGroupThresholdInvalid
	}
	if groupThreshold > len(groups) {
		return Spec{}, ErrGroupThresholdInvalid
	}
	if len(groups) > MaxShareCount {
		return Spec{}, ErrGroupCountInvalid
	}
	copied := append([]GroupSpec(nil), groups...)
	return Spec{groupThreshold: groupThreshold, groups: copied}, nil
}

// GroupThreshold returns the number of groups required for reconstruction.
func (s Spec) GroupThreshold() int {
	return s.groupThreshold
}

// Groups returns a copy of the group specifications.
func (s Spec) Groups() []GroupSpec {
	return append([]GroupSpec(nil), s.groups...)
}

// GroupCount returns the number of groups in the spec.
func (s Spec) GroupCount() int {
	return len(s.groups)
}

// ShareCount returns the total number of member shares across all groups.
func (s Spec) ShareCount() int {
	total := 0
	for _, group := range s.groups {
		total += group.MemberCount()
	}
	return total
}

// GroupSpec is a specification for one group of shares.
type GroupSpec struct {
	memberThreshold int
	memberCount     int
}

// NewGroupSpec creates a new GroupSpec with the given member threshold and count.
func NewGroupSpec(memberThreshold, memberCount int) (GroupSpec, error) {
	if memberCount == 0 {
		return GroupSpec{}, ErrMemberCountInvalid
	}
	if memberCount > MaxShareCount {
		return GroupSpec{}, ErrMemberCountInvalid
	}
	if memberThreshold > memberCount {
		return GroupSpec{}, ErrMemberThresholdInvalid
	}
	return GroupSpec{memberThreshold: memberThreshold, memberCount: memberCount}, nil
}

// MemberThreshold returns the member threshold for the group.
func (g GroupSpec) MemberThreshold() int {
	return g.memberThreshold
}

// MemberCount returns the number of members in the group.
func (g GroupSpec) MemberCount() int {
	return g.memberCount
}

// ParseGroupSpec parses a string in the form "<memberThreshold>-of-<memberCount>".
func ParseGroupSpec(s string) (GroupSpec, error) {
	parts := strings.Split(s, "-")
	if len(parts) != 3 {
		return GroupSpec{}, ErrGroupSpecInvalid
	}
	memberThreshold, err := strconv.Atoi(parts[0])
	if err != nil {
		return GroupSpec{}, ErrGroupSpecInvalid
	}
	if parts[1] != "of" {
		return GroupSpec{}, ErrGroupSpecInvalid
	}
	memberCount, err := strconv.Atoi(parts[2])
	if err != nil {
		return GroupSpec{}, ErrGroupSpecInvalid
	}
	return NewGroupSpec(memberThreshold, memberCount)
}

// DefaultGroupSpec returns the default 1-of-1 group spec.
func DefaultGroupSpec() GroupSpec {
	group, err := NewGroupSpec(1, 1)
	if err != nil {
		panic(err)
	}
	return group
}

// String renders the group spec as "<threshold>-of-<count>".
func (g GroupSpec) String() string {
	return fmt.Sprintf("%d-of-%d", g.memberThreshold, g.memberCount)
}
